packages/server/modules/gatekeeper/services/subscriptions/upgradeWorkspaceSubscription.ts

```diff
@@ -0,0 +1,381 @@
+      const guestProduct = subscriptionData.products.find(
+        (p) => p.productId === getWorkspacePlanProductId({ workspacePlan: 'guest' })
+      )
+      if (guestProduct) {
+        mutateSubscriptionDataWithNewValidSeatNumbers({
+          seatCount: 0,
+          getWorkspacePlanProductId,
+          subscriptionData,
+          workspacePlan: 'guest'
+        })
+
+        subscriptionData.products.push({
+          quantity: guestCount,
+          productId: getWorkspacePlanProductId({ workspacePlan: 'guest' }),
+          priceId: getWorkspacePlanPriceId({
+            workspacePlan: 'guest',
+            billingInterval
+          }),
+          subscriptionItemId: undefined
+        })
+      }
+    }
+
+    // set current plan seat count to 0
+    mutateSubscriptionDataWithNewValidSeatNumbers({
+      seatCount: 0,
+      getWorkspacePlanProductId,
+      subscriptionData,
+      workspacePlan: workspacePlan.name
+    })
+
+    // set target plan seat count to current seat count
+    subscriptionData.products.push({
+      quantity: memberCount + adminCount,
+      productId: getWorkspacePlanProductId({ workspacePlan: targetPlan }),
+      priceId: getWorkspacePlanPriceId({
+        workspacePlan: targetPlan,
+        billingInterval
+      }),
+      subscriptionItemId: undefined
+    })
+
+    await reconcileSubscriptionData({
+      subscriptionData,
+      prorationBehavior: isNewPlanType(targetPlan)
+        ? 'always_invoice'
+        : 'create_prorations'
+    })
+    await upsertWorkspacePlan({
+      workspacePlan: {
+        status: workspacePlan.status,
+        workspaceId,
+        name: targetPlan,
+        createdAt: new Date()
+      }
+    })
+    await updateWorkspaceSubscription({ workspaceSubscription })
+  }
+
+export const upgradeWorkspaceSubscriptionFactoryNew =
+  ({
+    getWorkspacePlan,
+    getWorkspacePlanProductId,
+    getWorkspacePlanPriceId,
+    getWorkspaceSubscription,
+    reconcileSubscriptionData,
+    updateWorkspaceSubscription,
+    countSeatsByTypeInWorkspace,
+    upsertWorkspacePlan
+  }: {
+    getWorkspacePlan: GetWorkspacePlan
+    getWorkspacePlanProductId: GetWorkspacePlanProductId
+    getWorkspacePlanPriceId: GetWorkspacePlanPriceId
+    getWorkspaceSubscription: GetWorkspaceSubscription
+    reconcileSubscriptionData: ReconcileSubscriptionData
+    updateWorkspaceSubscription: UpsertWorkspaceSubscription
+    countSeatsByTypeInWorkspace: CountSeatsByTypeInWorkspace
+    upsertWorkspacePlan: UpsertPaidWorkspacePlan
+  }) =>
+  async ({
+    workspaceId,
+    targetPlan,
+    billingInterval
+  }: {
+    workspaceId: string
+    targetPlan: PaidWorkspacePlans
+    billingInterval: WorkspacePlanBillingIntervals
+  }) => {
+    const workspacePlan = await getWorkspacePlan({
+      workspaceId
+    })
+
+    if (!workspacePlan) throw new WorkspacePlanNotFoundError()
+    if (!isNewPlanType(workspacePlan.name) || !isNewPlanType(targetPlan)) {
+      throw new NotImplementedError()
```

Comment:
In the case of `!isNewPlanType(targetPlan)` this is not a valid throw. It's not that its not implemented, its that its an explicitly unsupported operation and should show an appropriate error just like on line  120

The same issue is in the old implementation function on line 112